### PR TITLE
Fix iostream related issues

### PR DESCRIPTION
**Pull request by igrr**
* add stubs for more exception throw calls (https://github.com/esp8266/Arduino/issues/3358)
* make putc_r implementation weak (https://github.com/esp8266/Arduino/issues/4630)
* fix incorrect return value of _write_r call
* add device test for output to std::cout

### CODE REVIEW

File: tests/device/test_iostream/test_iostream.ino [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    Serial.begin(115200);
+    BS_RUN(Serial);
+}
+
+TEST_CASE("can print to std::cout", "[iostream]")
+{
+    std::stringstream test_stream("");
+    test_stream << "hello stream";
+
+	Serial.readString();
```
Comment (d-a-v) [inline tests/device/test_iostream/test_iostream.ino]: Why is this needed ?

Comment (devyte) [reply tests/device/test_iostream/test_iostream.ino]: Same comment as @d-a-v

File: cores/esp8266/libc_replacements.c [hunk 2]
```diff
@@ -84,15 +84,18 @@ int ICACHE_RAM_ATTR _read_r(struct _reent* unused, int file, char *ptr, int len)
 
 int ICACHE_RAM_ATTR _write_r(struct _reent* r, int file, char *ptr, int len) {
     (void) r;
+    int pos = len;
     if (file == STDOUT_FILENO) {
-        while(len--) {
+        while(pos--) {
```
Comment (devyte) [inline cores/esp8266/libc_replacements.c]: Does this address a current issue or ongoing discussion? Maybe a link could be added in the PR description if that's the case.

File: tests/device/test_iostream/test_iostream.ino (hunk 1 above)
Comment (igrr) [reply tests/device/test_iostream/test_iostream.ino]: Was meant to clean up UART RX buffer just in case there was something in there. Chances are, the test will also work without this.

And I have messed up the indentation in this line...

File: cores/esp8266/libc_replacements.c (hunk 2 above)
Comment (igrr) [reply cores/esp8266/libc_replacements.c]: Haven't seen this reported, probably because including iostream header caused compilation error, so no one got as far as using iostreams. Found when writing the test case.
